### PR TITLE
Add finding best suppliers, clients

### PR DESCRIPTION
**Pull request by rostIvan**

### CODE REVIEW

File: src/main/java/com/ita/if103java/ims/dao/impl/BestTradeDaoImpl.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+            return jdbcTemplate.query(
+                Queries.SQL_SELECT_BEST_TOP_WAREHOUSES_BY_ITEM_ID,
+                (rs, i) -> new TotalTransactionQuantity(
+                    rs.getLong("top_warehouse__id"),
+                    rs.getLong("total")
+                ),
+                itemId, limit
+            );
+        } catch (DataAccessException e) {
+            throw new CRUDException("Error during `select` best associates by item_id", e);
```
Comment (ikrasWork) [inline src/main/java/com/ita/if103java/ims/dao/impl/BestTradeDaoImpl.java]: fix message

File: src/main/java/com/ita/if103java/ims/dao/impl/BestTradeDaoImpl.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+                itemId, type.toString(), limit
+            );
+        } catch (DataAccessException e) {
+            throw new CRUDException("Error during `select` best associates by item_id", e);
+        }
+    }
+
+    public static final class Queries {
+        public static final String SQL_SELECT_BEST_TOP_WAREHOUSES_BY_ITEM_ID = """
+                select IFNULL(w1.top_warehouse_id, w2.top_warehouse_id) as `top_warehouse__id`,
```
Comment (ikrasWork) [inline src/main/java/com/ita/if103java/ims/dao/impl/BestTradeDaoImpl.java]: for warehouse we need to select all warehouses and do not foun top warehouses

File: src/main/java/com/ita/if103java/ims/dao/impl/BestTradeDaoImpl.java (hunk 1 above)
Comment (rostIvan) [reply src/main/java/com/ita/if103java/ims/dao/impl/BestTradeDaoImpl.java]: Oops

File: src/main/java/com/ita/if103java/ims/dao/impl/BestTradeDaoImpl.java (hunk 2 above)
Comment (rostIvan) [reply src/main/java/com/ita/if103java/ims/dao/impl/BestTradeDaoImpl.java]: How about this?
```sql
from transactions t
         left join warehouses w1 on t.moved_from = w1.id
         left join warehouses w2 on t.moved_to = w2.id
where t.item_id = ?
  ...
  and (t.moved_from != w1.top_warehouse_id or t.moved_to != w2.top_warehouse_id)
```

Comment (ikrasWork) [reply src/main/java/com/ita/if103java/ims/dao/impl/BestTradeDaoImpl.java]: simpler
```sql
SELECT * FROM warehouses
WHERE parent_id IS NULL AND active = 1;
```

Comment (rostIvan) [reply src/main/java/com/ita/if103java/ims/dao/impl/BestTradeDaoImpl.java]: Ok, got it.
Thanks a lot

Comment (rostIvan) [reply src/main/java/com/ita/if103java/ims/dao/impl/BestTradeDaoImpl.java]: Better to add it in new PR.
In WarehouseDao for e.g.

**Pull request merged**